### PR TITLE
Fix GROUP BY AggregateFunction

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `GROUP BY` `AggregateFunction` (i.e. you `GROUP BY` by the column that has `AggregateFunction` type)

### CODE REVIEW

**Comment by azat** [comment]
>[Stateless tests (thread, actions) [3/3] — fail: 1, passed: 1336, skipped: 7](https://s3.amazonaws.com/clickhouse-test-reports/37093/2278faeb6d2e3b93d66f396742ab484da0ced591/stateless_tests__thread__actions__[3/3].html)

- `01119_optimize_trivial_insert_select` - #37101

File: src/Processors/Transforms/TotalsHavingTransform.cpp [hunk 1]
```diff
@@ -17,33 +17,39 @@ namespace ErrorCodes
...
+static void finalizeBlock(Block & block, const ColumnNumbers & aggregation_keys)
 {
     for (size_t i = 0; i < block.columns(); ++i)
     {
+        if (std::find(aggregation_keys.begin(), aggregation_keys.end(), i) != aggregation_keys.end())
+            continue;
+
         ColumnWithTypeAndName & current = block.getByPosition(i);
         const DataTypeAggregateFunction * unfinalized_type = typeid_cast<const DataTypeAggregateFunction *>(current.type.get());
+        assert(unfinalized_type);
```
Comment (nickitat) [inline src/Processors/Transforms/TotalsHavingTransform.cpp:46]: maybe cast to a reference and get an exception if smth went wrong?

Comment (azat) [reply src/Processors/Transforms/TotalsHavingTransform.cpp]: Good suggestion, done, thanks.

**Review by nickitat — APPROVED** [review]

**Comment by azat** [comment]
>Conflicting files
>src/Interpreters/InterpreterSelectQuery.cpp

Fixed.

**Comment by nickitat** [comment]
stress tests: https://github.com/ClickHouse/ClickHouse/issues/37094

**Comment by azat** [comment]
> this is likely caused by this pr: https://s3.amazonaws.com/clickhouse-test-reports/37093/ed2ea9f327967771dc37d3e72b4e9f26592e9b32/stress_test__undefined__actions_/fatal_messages.txt

Indeed, positions was incorrect, fixed.

File: src/Processors/Transforms/finalizeChunk.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <Interpreters/AggregateDescription.h>
+#include <Core/Block.h>
+#include <Processors/Chunk.h>
+#include <vector>
+
+namespace DB
+{
+
+using ColumnsByPosition = std::vector<bool>;
```
Comment (nickitat) [inline src/Processors/Transforms/finalizeChunk.h:11]: this name doesn't really fit here IMO. what about `ColumnsMask` or smth similar? or get back to just a list of indices?

Comment (azat) [reply src/Processors/Transforms/finalizeChunk.h]: Agree, ColumnsMask looks better. Rebased.

**Review by nickitat — APPROVED** [review]

**Comment by azat** [comment]
>[Stress test (address, actions) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt)]

```
2022.05.16 18:10:37.009571 [ 672692 ] {} <Error> void DB::StorageWindowView::threadFuncFireProc(): Code: 20. DB::Exception: Number of columns doesn't match. (NUMBER_OF_COLUMNS_DOESNT_MATCH), Stack trace (when copying this message, always include the lines below):
```

>[Stress test (undefined, actions) — Server failed to start (see application_errors.txt and clickhouse-server.clean.log)](https://s3.amazonaws.com/clickhouse-test-reports/37093/05dc95d849ce17bdca62226df40803512e9944ec/stress_test__undefined__actions_.html)

```
2022.05.16 23:45:45.463330 [ 85886 ] {} <Error> Application: DB::Exception: The local set of parts of table test_p697zd.trunc doesn't look like the set of parts in ZooKeeper: 6.00 rows of 6.00 total rows in filesystem are suspicious. There are 3 unexpected parts with 6 rows (0 of them is not just-written with 0 rows), 0 missing parts (with 0 blocks).: Cannot attach table `test_p697zd`.`trunc` from metadata file /var/lib/clickhouse/metadata/test_p697zd/trunc.sql from query ATTACH TABLE test_p697zd.trunc (`n` Int32) ENGINE = ReplicatedMergeTree('/test/1166/test_p697zd', '1') PARTITION BY n % 10 PRIMARY KEY n ORDER BY n SETTINGS index_granularity = 8192
```

**Pull request merged**